server/pulp/server/async/scheduler.py

```diff
@@ -317,15 +320,60 @@ def tick(self):
         :return:    number of seconds before the next tick should run
         :rtype:     float
         """
-        ret = super(Scheduler, self).tick()
         self._failure_watcher.trim()
-
         # this is not an event that gets sent anywhere. We process it
         # immediately.
         scheduler_event = {'timestamp': time.time(),
                            'type': 'scheduler-event',
-                           'hostname': ("%s@%s" % (SCHEDULER_WORKER_NAME, platform.node()))}
+                           'hostname': ("%s@%s@%s" % (SCHEDULER_WORKER_NAME,
+                                                      str(celery_name), platform.node()))}
         worker_watcher.handle_worker_heartbeat(scheduler_event)
+
+        celerybeat_name = SCHEDULER_WORKER_NAME + "@" + str(celery_name) + "@" + platform.node()
+
+        # Establishing connection with DB
+        db = db_connection.get_database()
+        collection = db['celerybeat_lock']
+
+        # Creating a unique constraint on the collection
+        collection.create_index('lock', unique=True)
+
+        _logger.debug('Updating the lock with celerybeat_name %s ', celerybeat_name)
+
+        old_timestamp = datetime.utcnow() - timedelta(seconds=CELERYBEAT_OLD_TIMESTAMP)
+
+        # Updating the current lock if lock is on this instance of celerybeat
+        result = collection.find_and_modify(
+            query={'celerybeat_name': celerybeat_name},
+            update={'$set': {'lock_timestamp': datetime.utcnow()}}
+        )
+
+        # If current instance has lock and updated lock_timestamp, call super
+        if result is not None:
+            ret = super(Scheduler, self).tick()
+            _logger.info("Lock updated, Ticking again in %s secs "
+                         "and celerybeat_name is %s ", ret, celerybeat_name)
+        else:
+            # check for old enough time_stamp and remove if such lock is present
+            collection.remove({'lock_timestamp': {'$lte': old_timestamp}})
+            try:
+                # Insert new lock entry
+                collection.insert({
+                    'celerybeat_name': celerybeat_name,
+                    'lock_timestamp': datetime.utcnow(),
+                    'lock': "locked"
+                })
+                _logger.info("New lock acquired by %s ", celerybeat_name)
+                # After acquiring new lock call super to dispatch tasks
+                ret = super(Scheduler, self).tick()
+                _logger.info("Ticking again in %s secs and "
```

Comment:
This should be internationalized with _() and you need to switch the logging to dictionary based (not position based).